library/ssl_tls13_server.c

```diff
@@ -2780,9 +2842,102 @@ static int ssl_tls13_write_server_finished(mbedtls_ssl_context *ssl)
         mbedtls_ssl_handshake_set_state(ssl, MBEDTLS_SSL_CLIENT_FINISHED);
     }
 
+    MBEDTLS_SSL_DEBUG_MSG(2, ("<= ssl_tls13_process_wait_flight2"));
     return 0;
 }
 
+#if defined(MBEDTLS_SSL_EARLY_DATA)
+/*
+ * Handler for MBEDTLS_SSL_END_OF_EARLY_DATA( WAIT_EOED )
+ *
+ * RFC 8446 section A.2
+ *
+ *                 |
+ *    +------> WAIT_EOED -+
+ *    |       Recv |      | Recv EndOfEarlyData
+ *    | early data |      | K_recv = handshake
+ *    +------------+      |
+ *                        |
+ *  WAIT_FLIGHT2 <--------+
+ *                                 |
+ */
+MBEDTLS_CHECK_RETURN_CRITICAL
+static int ssl_tls13_process_wait_eoed(mbedtls_ssl_context *ssl)
+{
+    int ret = MBEDTLS_ERR_ERROR_CORRUPTION_DETECTED;
+    mbedtls_ssl_handshake_params *handshake = ssl->handshake;
+
+    MBEDTLS_SSL_DEBUG_MSG(2, ("=> ssl_tls13_process_wait_eoed"));
+
+    if ((ret = mbedtls_ssl_read_record(ssl, 0)) != 0) {
```

Comment:
Usually, we use mbedtls_ssl_tls13_fetch_handshake_msg() to read messages instead of mbedtls_ssl_read_record()

Reply:
Then, should we wrap mbedtls_ssl_read_record in ssl_tls13_generic.c, and then call the wrapper function?  Actually we have had the similar wrapper function, should we align them to avoid calling into the api in ssl_msg.c to break the independency?

Reply:
> I do not found that, could you point it? In fact , `ssl_tls13_certificate_request_coordinate()` does similar thing also.

See mbedtls_ssl_tls13_fetch_handshake_msg() and mbedtls_ssl_tls13_write_change_cipher_spec().
Anyway, I am fine to use it as you have done if somewhere has done the similar things.

Reply:
If no other points, pleas close this conversation. It really bother me